python/dgl/graph.py

```diff
@@ -188,10 +188,12 @@ def __init__(self,
             self._edge_frame = FrameRef(Frame(num_rows=self.number_of_edges()))
         else:
             self._edge_frame = edge_frame
-        # msg graph & frame
-        self._msg_graph = create_graph_index(multigraph=multigraph)
-        self._msg_frame = FrameRef()
-        self.reset_messages()
+        self._msg_index = utils.zero_index(size=self.number_of_edges())
+        # message frame
+        self._msg_frame = FrameRef(Frame(num_rows=self.number_of_edges()))
+        # set initializer for message frame
+        # all fields including message indicator should use zero_initializer
+        self._msg_frame.set_initializer(dgl.init.zero_initializer)
```

Comment:
Would zero initialization impact non-sum reducers (e.g. max)?  Especially if we're doing builtin max with kernel fusion in the future.

Reply:
This has nothing to do with reduce. msg_frame only stores message, and whether message exists is indicated by `graph._msg_index`, which is a boolean index of size (num_edges). Each element represents whether there is a message on this edge.

And messages filled in by zero_initializer won't be used because the program always check `graph._msg_index` first. In fact, you can replace zero_initializer with anything...

Reply:
NVM I didn't realize it is only being used in `send`, also I missed the schedule recv part...